src/gui/mainwindow.cpp

```diff
@@ -1915,21 +1918,18 @@ void MainWindow::handleUpdateCheckFinished(const bool updateAvailable, const QSt
         msgBox->setAttribute(Qt::WA_DeleteOnClose);
         msgBox->setAttribute(Qt::WA_ShowWithoutActivating);
         msgBox->setDefaultButton(QMessageBox::Yes);
-        connect(msgBox, &QMessageBox::buttonClicked, this, [this, msgBox, signalSender](QAbstractButton *button)
+        connect(msgBox, &QMessageBox::buttonClicked, this, [this, msgBox, updater](QAbstractButton *button)
         {
             if (msgBox->buttonRole(button) == QMessageBox::YesRole)
             {
-                // The user want to update, let's download the update
-                auto *updater = dynamic_cast<ProgramUpdater *>(signalSender);
                 updater->updateProgram();
             }
-            else
-            {
-                if (Preferences::instance()->isUpdateCheckEnabled())
-                    m_programUpdateTimer->start();
-            }
-
-            signalSender->deleteLater();
+        });
+        connect(msgBox, &QDialog::finished, this, [this, updater](const int)
+        {
+            if (m_programUpdateTimer)
+                m_programUpdateTimer->start();
+            updater->deleteLater();
```

Comment:
I'm sorry, but the reasons why ProgramUpdater instance can't be reused multiple times (so you delete it and re-create it next time) are not obvious to me. However, I won't bother you with this issue, if you think the current logic is digestible.

Reply:
The reason is due to my previous comment: https://github.com/qbittorrent/qBittorrent/pull/14274#discussion_r563702989
In short, either deal with `connect()`, `disconnect()` or init a new instance (to omit `disconnect()`) every time.

Reply:
Okay, that's not so important right now. But it looks like you won't be able to implement any asynchronous logic in `ProgramUpdater::updateProgram()` if you need it in the future.